sdk/typespec/typespec_client_core/src/http/request/mod.rs

```diff
@@ -217,6 +217,26 @@ impl<T> From<Body> for RequestContent<T> {
     }
 }
 
+#[cfg(all(feature = "tokio_fs", not(target_arch = "wasm32")))]
+impl<T> From<crate::fs::FileStream> for RequestContent<T> {
+    fn from(body: crate::fs::FileStream) -> Self {
+        Self {
+            body: body.into(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[cfg(all(feature = "tokio_fs", not(target_arch = "wasm32")))]
+impl<T> From<&crate::fs::FileStream> for RequestContent<T> {
+    fn from(body: &crate::fs::FileStream) -> Self {
+        Self {
+            body: body.into(),
+            phantom: PhantomData,
+        }
+    }
+}
+
```

Comment:
@RickWinter @JeffreyRichter @analogrelay @demoray @cataggar what do you think? Think this may be common enough for request data streaming to add first-tier support for streaming a file directly into `RequestContent`? I'm leaning that way, but maybe 60/40 or so. I generally agree with "expose things as you need them" and we certainly have several ways to get binary data - even streaming (well, one way) - into a request, but I think this greatly simplifies the API for a common case *for some* services e.g., Storage (or anything that can take a file upload directly even if it ends up in Storage).

Reply:
I agree with you on the "expose things as you need them" and so I'm hesitant to add this now.
In the early Rust days, we had streaming, and we took it out - I remember Yosh was very happy when we did this because it is complicated when you also consider retries and having to rewind/seek the stream back to the beginning.

Reply:
But because of the orphan rule, crates that do need it - say, `azure_storage_blobs` - couldn't add this themselves if they want to support streaming a large file in a request.

Reply:
What we did in other languages is take a generic stream (file, network or something else). The stream must be readable and seekable (for retries). Is there a generic stream in Rust standard library we could use?

Reply:
In `std`, no. In fact, there's hardly anything in `std`. But we do declare our own `SeekableStream`. `Body` already has a `From<Box<dyn SeekableStream>>` so we could add one here. It still eliminates a line but callers will have to box the stream. It's not so elegant:

```rust
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = FileStream::open("foo.txt")?;
    let bytes = read(Box::new(file))?;

    println!("read {} bytes", bytes.len());

    Ok(())
}
```

Still, not so bad either.

Reply:
Actually, this is even uglier:

```rust
        let file = fs::File::open(file!()).await?;
        let file: Box<dyn SeekableStream> = Box::new(
            FileStreamBuilder::new(file)
                // Simulate a slow, chunky request.
                .buffer_size(256usize)
                .build()
                .await?,
        );
        client::put_binary_data(file.into()).await?;
```

I think we should keep the explicit `From<FileStream>` and `From<&FileStream>` but can colocate them into `tokio.fs` like the ones for `Body` are now.